api/share/utils.py

```diff
@@ -0,0 +1,363 @@
+        GraphNode('throughtags', creative_work=registration_graph, tag=GraphNode('tag', name=tag._id))
+        for tag in registration.tags.all() or [] if tag._id
+    ]
+
+    to_visit.extend(format_contributor(registration_graph, user, bool(user._id in registration.visible_contributor_ids), i) for i, user in enumerate(registration.contributors))
+    to_visit.extend(GraphNode('AgentWorkRelation', creative_work=registration_graph, agent=GraphNode('institution', name=institution.name)) for institution in registration.affiliated_institutions.all())
+
+    if registration.parent_node:
+        parent = GraphNode('registration')
+        to_visit.extend([
+            parent,
+            GraphNode('workidentifier', creative_work=parent, uri=urljoin(settings.DOMAIN, registration.parent_node.url)),
+            GraphNode('ispartof', subject=registration_graph, related=parent),
+        ])
+
+    visited = set()
+    to_visit.extend(registration_graph.get_related())
+
+    while True:
+        if not to_visit:
+            break
+        n = to_visit.pop(0)
+        if n in visited:
+            continue
+        visited.add(n)
+        to_visit.extend(list(n.get_related()))
+
+    return [node_.serialize() for node_ in visited]
+
+
+def send_desk_share_error(resource, resp, retries):
+    from osf.models import Preprint
+
+    is_preprint = isinstance(resource, Preprint)
+
+    if is_preprint:
+        kwargs = {
+            'logo': settings.OSF_PREPRINTS_LOGO,
+            'preprint': resource,
+        }
+    else:
+        kwargs = {'node': resource}
+
+    mails.send_mail(
+        to_addr=settings.OSF_SUPPORT_EMAIL,
+        mail=mails.SHARE_PREPRINT_ERROR_DESK if is_preprint else mails.SHARE_ERROR_DESK,
+        resp=resp,
+        retries=retries,
+        can_change_preferences=False,
+        **kwargs
+    )
+
+
+def is_qa_resource(resource):
+    """
+    QA puts tags and special titles on their project to stop them from appearing in the search results. This function
+    check if a resource is a 'QA resource' that should be indexed.
+    :param resource: should be Node/Registration/Preprint
+    :return:
+    """
+    tags = set(resource.tags.all().values_list('name', flat=True))
+    has_qa_tags = bool(set(settings.DO_NOT_INDEX_LIST['tags']).intersection(tags))
+
+    has_qa_title = any(substring in resource.title for substring in settings.DO_NOT_INDEX_LIST['titles'])
+
+    return has_qa_tags or has_qa_title
+
+
+def update_share(resource, old_subjects=None):
+    data = serialize_share_data(resource, old_subjects)
+    resp = send_share_json(resource, data)
+    try:
+        resp.raise_for_status()
+    except requests.HTTPError:
+        if resp.status_code >= 500:
+            async_update_resource_share.delay(resource._id, old_subjects)
+        else:
+            send_desk_share_error(resource, resp, 0)
+
+@celery_app.task(bind=True, max_retries=4, acks_late=True)
+def async_update_resource_share(self, guid, old_subjects=None):
```

Comment:
Consolidating `async_update_node_share` and `async_update_prepriny_share`, might seem like a little scope creepy, but when we are asynchronously making external requests we really want to cut down on the number of moving parts and redundancies, especially because of the mocking and coverage issues present in the ticket.